src/nexmark/mod.rs

```diff
@@ -36,6 +36,72 @@ pub mod generator;
 pub mod model;
 pub mod queries;
 
+/// BatchedReceiver abstracts the Receiver interface for channels of VecDeque's.
+pub struct BatchedReceiver<T> {
+    rx: mpsc::Receiver<VecDeque<T>>,
+    current_queue: VecDeque<T>,
+}
+
+impl<T> BatchedReceiver<T> {
+    fn new(rx: mpsc::Receiver<VecDeque<T>>) -> Self {
+        Self {
+            rx,
+            current_queue: VecDeque::new(),
+        }
+    }
+
+    fn recv(&mut self) -> Result<T, mpsc::RecvError> {
+        match self.current_queue.pop_front() {
+            Some(t) => Ok(t),
+            None => {
+                self.current_queue = self.rx.recv()?;
+                Ok(self.current_queue.pop_front().unwrap())
+            }
+        }
+    }
+}
+
+/// BatchedSender abstracts the Sender interface for channels of VecDeque's.
+pub struct BatchedSender<T> {
+    batch_size: usize,
+    current_queue: VecDeque<T>,
+    tx: mpsc::SyncSender<VecDeque<T>>,
+}
+
+fn batched_channel<T>(batch_size: usize) -> (BatchedSender<T>, BatchedReceiver<T>) {
+    let (tx, rx) = mpsc::sync_channel(3);
+    (
+        BatchedSender {
+            batch_size,
+            current_queue: VecDeque::<T>::with_capacity(batch_size),
+            tx,
+        },
+        BatchedReceiver::new(rx),
+    )
+}
+
+impl<T> BatchedSender<T> {
+    pub fn send(&mut self, t: T) -> Result<(), mpsc::SendError<VecDeque<T>>> {
+        // The order here ensures that we never end up sending an empty
+        // vector.
+        if self.current_queue.len() == self.batch_size {
+            self.tx.send(std::mem::replace(
```

Comment:
I'd not seen the use of `std::mem::replace` like this to be able to get ownership of a field in an `&mut self` to use, while setting the field to a new value... neat. Lucky I found [a relevant post](https://smallcultfollowing.com/babysteps/blog/2018/11/10/after-nll-moving-from-borrowed-data-and-the-sentinel-pattern/) or I would have struggled for a while.